services/src/main/scala/cromwell/pipeline/service/GitLabProjectVersioning.scala

```diff
@@ -21,44 +20,57 @@ class GitLabProjectVersioning(httpClient: HttpClient, config: GitLabConfig)
     val versionValue = version.map(_.name).getOrElse(config.defaultFileVersion)
 
     httpClient
-      .put(
+      .put[Message, UpdateFileRequest](
         fileUrl,
-        payload = Json.stringify(
-          Json.toJson(UpdateFileRequest(versionValue, projectFile.content, versionValue))
-        ),
+        payload = UpdateFileRequest(versionValue, projectFile.content, versionValue),
         headers = config.token
       )
       .flatMap {
         case Response(HttpStatusCodes.OK, _, _) =>
           Future.successful(Right("Success update file"))
         case Response(HttpStatusCodes.BadRequest, _, _) =>
           httpClient
-            .post(
+            .post[Message, UpdateFileRequest](
               fileUrl,
-              payload = Json.stringify(
-                Json.toJson(UpdateFileRequest(config.defaultFileVersion, projectFile.content, "Init commit"))
-              ),
+              payload = UpdateFileRequest(config.defaultFileVersion, projectFile.content, "Init commit"),
               headers = config.token
             )
             .map {
-              case Response(HttpStatusCodes.OK, _, _) => Right("Create new file")
-              case Response(_, body, _)               => Left(VersioningException(body))
+              case Response(HttpStatusCodes.OK, _, _)         => Right("Create new file")
+              case Response(_, NegativeResponseBody(body), _) => Left(VersioningException(body))
+              case _ =>
+                Left(
+                  VersioningException(
+                    "status is not Ok, but response is not negative, check status and possible positive responses"
```

Comment:
The existence of this line indicates that something went wrong during the implementation of the HTTP client.
Separating bad and good responses is a good idea. But you must make sure that whoever uses your client won't have to assume that something went wrong in the HTTP client implementation.

Reply:
As I remember this was made just in order to make the match exhaustive. For the cases where we try to make a request and get unexpected success response(which have not already exist in httpclient match)